src/librustc_typeck/check/never_compat.rs

```diff
@@ -0,0 +1,433 @@
+                            debug!(
+                                "Type variable {:?} is equal to diverging var {:?}",
+                                var, diverging_var
+                            );
+
+                            debug!(
+                                "Var origin: {:?}",
+                                fcx.infcx.type_variables.borrow().var_origin(*vid1)
+                            );
+                            return VarData { best_var: *vid1, best_diverging_var: *vid2 };
+                        }
+                    }
+                    _ => bug!("Unexpected types: var={:?} diverging_var={:?}", var, diverging_var),
+                }
+            }
+        }
+        bug!("No vars were equated to diverging vars: {:?}", vars)
+    }
+
+    /// The hook used by typecheck after fallback has been run.
+    /// We look for any "questionable" calls, generating diagnostic
+    /// message for them.
+    pub fn post_fallback(self, fcx: &FnCtxt<'a, 'tcx>) {
+        let tcx = fcx.tcx;
+        for (call_id, path) in &self.unresolved_paths {
+            debug!(
+                "post_fallback: resolved ty: {:?} at span {:?} : expr={:?} parent={:?} path={:?}",
+                path.span,
+                path.ty,
+                tcx.hir().get(*call_id),
+                tcx.hir().get(tcx.hir().get_parent_node(*call_id)),
+                path
+            );
+
+            let span = path.span;
+            if let Some(vars) = find_questionable_call(path, fcx) {
+                let VarData { best_var, best_diverging_var } = self.find_best_vars(fcx, &vars);
+
+                let var_origin = *fcx.infcx.type_variables.borrow().var_origin(best_var);
+                let diverging_var_span =
+                    fcx.infcx.type_variables.borrow().var_origin(best_diverging_var).span;
+
+                let mut err = tcx
+                    .sess
+                    .struct_span_err(span, "Fallback to `!` may introduce undefined behavior");
+
+                // There are two possible cases here:
+                match var_origin.kind {
+                    // 1. This inference variable was automatically generated due to the user
+                    // not using the turbofish syntax. For example, `foo()` where foo is deinfed
+                    // as `fn foo<T>() { ... }`.
+                    // In this case, we point at the definition of the type variable (e.g. the `T`
+                    // in `foo<T>`), to better explain to the user what's going on.
+                    TypeVariableOriginKind::TypeParameterDefinition(name, did) => {
+                        err.span_note(
+                            var_origin.span,
+                            &format!("the type parameter {} here was inferred to `!`", name),
+                        );
+                        if let Some(did) = did {
+                            err.span_note(fcx.tcx.def_span(did), "(type parameter defined here)");
```

Comment:
This message ends up triggering when the 'type parameter' is `Self` - I think we should probably skip emitting the message instead.